### PR TITLE
Models

### PR DESCRIPTION
**Pull request by mweiss17**
This PR moves the risk prediction and message protocol implementation into the models directory. It is called on the output of the simulator. If you run this with the arg `--save_training_data` then it will output the risk predictions. This output can then be loaded into a super basic pytorch dataloader without batching.

Most of the documentation for this folder is contained in `docs/models.md`.

### CODE REVIEW

File: docs/models.md [hunk 1]
```diff
@@ -0,0 +1,77 @@
+# Risk Prediction, Message Passing, and Clustering
+
+## Overview
+This document describes the code in models. By running `python models/run.py`, you will 
+initiate a message passing algorithm which iterates over the logs output by the mobility simulator. We provide more details
+in the message passing section below. Those logs are described more in `events.md`, but briefly there are four kinds: encounters, symptom_start, contamination,
+and recovery. Our goal is to pass messages between people such that each person can accurately predict their own
```
Comment (satyaog) [inline docs/models.md]: there are five kinds: encounters, test, symptom_start, contamination, and recovery

File: docs/models.md [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+in the message passing section below. Those logs are described more in `events.md`, but briefly there are four kinds: encounters, symptom_start, contamination,
+and recovery. Our goal is to pass messages between people such that each person can accurately predict their own 
+risk of being infectious while maintaining the highest level of privacy. 
+
+## Privacy
+* Risk values are quantized to 4-bit codes such that they can take on one of 16 levels (0-15). 
+* User ids are also quantized to 4-bit codes into and left-shifted each day, with a new random bit appended to the right. 
+
+## Message Passing
+When you run `python models/run.py`, we load the logs contained in `output/data.pkl`, then initialize the people in the sim:
```
Comment (satyaog) [inline docs/models.md]: `output/data.zip` is the new file

File: models/run.py [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+parser.add_argument('--save_training_data', action="store_true")
+
+
+def main(args):
+    # Define constants
+    SIGNIFICANT_RISK_LEVEL_CHANGE = 0.0625
+
+    # read and filter the pickles
+    with open(args.data_path, "rb") as f:
+        logs = pickle.load(f)
```
Comment (satyaog) [inline models/run.py]: This is to load pickles from data.zip

```
logs = []
with zipfile.ZipFile(args.data_path, 'r') as zf:
    logs.extend([pickle.load(zf.open(pkl, 'r')) for pkl in zf.namelist()])

# Sort the logs
logs.sort(key=lambda e: (int(e['human_id'][6:]), e['time']))
```

**Review by satyaog — CHANGES_REQUESTED** [review]
The events are now logged as pickles in a zip file. The review suggests a method to load them but it's important to note that the iterative logs in the zip was to reduce the memory usage to fit in RAM. Loading all the pickles at once will not be possible for a zip produced by a simulation with lots of people

File: models/risk_models.py [hunk 4]
```diff
@@ -45,38 +51,46 @@ def update_risk_local(cls, human, now):
...
-        m_i_enc = _encode_message(m_i)
-        m_risk = binary_to_float("".join([str(x) for x in np.array(m_i[1].tolist()).astype(int)]), 0, 4)
-
-        # otherwise score against previous messages
+    def score_matches(cls, human, m_i):
         scores = {}
-        for m_enc, _ in human.M.items():
-            m = _decode_message(m_enc)
-            if m_i[0] == m[0] and m_i[2].day == m[2].day:
+        for m_enc, _ in reversed(human.M.items()):
```
Comment (satyaog) [inline models/risk_models.py]: It seams `__reverse__` on `dict_items` was only introduced in Python 3.8. Consider using `for m_enc, _ in reversed(list(human.M.items()))` for backward compatibility to at least Python 3.6

File: models/run.py [hunk 5]
```diff
@@ -3,40 +3,132 @@
...
+import argparse
 import subprocess
 import numpy as np
 import operator
+import datetime
+import time
+from tqdm import tqdm
 from collections import defaultdict
+from base import Event
 from dummy_human import DummyHuman
```
Comment (satyaog) [inline models/run.py]: `from models.dummy_human import DummyHuman` will allow running tests from the root run.py

File: models/run.py [hunk 6]
```diff
@@ -3,40 +3,132 @@
...
-from risk_models import RiskModelYoshua, RiskModelLenka, RiskModelEilif
-import datetime
+from risk_models import RiskModelYoshua, RiskModelLenka, RiskModelEilif, RiskModelTristan
```
Comment (satyaog) [inline models/run.py]: `from models.risk_models ...`

File: models/run.py [hunk 7]
```diff
@@ -3,40 +3,132 @@
...
 from plots.plot_risk import dist_plot, hist_plot
+from helper import messages_to_np, symptoms_to_np, candidate_exposures, rolling_infectiousness
```
Comment (satyaog) [inline models/run.py]: `from models.helper ...`

File: models/risk_models.py [hunk 8]
```diff
@@ -125,7 +139,62 @@ def update_risk_encounter(cls, human, message):
         human.M[msg_enc]['carry_over_transmission_proba'] = RISK_TRANSMISSION_PROBA * (1 - update)
 
 
+class RiskModelTristan(RiskModelBase):
+    risk_map = np.load("models/log_risk_mapping.npy")
```
Comment (satyaog) [inline models/risk_models.py]: `risk_map = np.load(f"{os.path.dirname(os.path.realpath(__file__))}/log_risk_mapping.npy")`

not sure if it is cleaner but at least it allows to start the preproc in a unittest runner which has its own working dir

File: models/risk_models.py [hunk 9]
```diff
@@ -6,18 +6,24 @@
 from utils import _encode_message, _decode_message, binary_to_float
```
Comment (satyaog) [inline models/risk_models.py:6]: Not sure if those lines are needed but they will not work in a unittest runner which has it's own working dir
```
import sys
import os
sys.path.append(os.getcwd())
```

File: models/run.py [hunk 10]
```diff
@@ -3,40 +3,132 @@
...
+                                            "is_recovered": is_recovered,
+                                            "recovery_day": recovery_day,
+                                            "exposed_locs": exposed_locs,
+                                            "exposure_encounter": exposure_encounter,
+                                            "infectiousness": infectiousness,
+                                        }
+                                    }
+    return {human.name: daily_output, "human": human}
+
+def main(args):
```
Comment (satyaog) [inline models/run.py]: Allows to run the code in a unittest
```
def main(args=None)
    args = args || parser.parse_args()
```

File: models/run.py [hunk 11]
```diff
@@ -48,96 +140,130 @@
...
-            clusters.append(human.M)
-        json.dump(clusters, open(CLUSTER_PATH, 'w'))
+    clusters = []
+    for human in hd.values():
+        clusters.append(human.M)
+    json.dump(clusters, open(args.cluster_path, 'w'))
 
+
+if __name__ == "__main__":
+    args = parser.parse_args()
```
Comment (satyaog) [inline models/run.py]: This can be moved inside main()

File: models/run.py [hunk 12]
```diff
@@ -3,40 +3,132 @@
...
-    CLUSTER_PATH = "output/clusters.json"
-    PATH_TO_PLOT = "plots/risk/"
-    RISK_MODEL = 'yoshua'  # options: ['yoshua', 'lenka', 'eilif']
-    if RISK_MODEL == 'yoshua':
-        RiskModel = RiskModelYoshua
-    LOGS_SUBSET_SIZE = 10000000
-    seed = 0
 
-    rng = np.random.RandomState(seed)
+def proc_human(params):
```
Comment (satyaog) [inline models/run.py:41]: `def proc_human(args, params)`

File: models/run.py [hunk 13]
```diff
@@ -48,96 +140,130 @@
...
+                    encountered_human = hd[unobs['human2']['human_id']]
+                    message = encountered_human.cur_message(current_day, RiskModel)
+                    encountered_human.sent_messages[
+                        str(unobs['human1']['human_id']) + "_" + str(encounter_time)] = message
+                    human.messages.append(message)
+                    got_exposed = encounter['payload']['unobserved']['human1']['got_exposed']
+                    if got_exposed:
+                        human.exposure_message = _encode_message(message)
+
+            daily_output = parallel((delayed(proc_human)(params) for params in all_params))
```
Comment (satyaog) [inline models/run.py:233]: `daily_output = parallel((delayed(proc_human)(args, params) for params in all_params))`

**Review by satyaog — CHANGES_REQUESTED** [review]
I think only the `reverse(list())` is important before a merge. The others can be added at a later time

File: docs/models.md (hunk 1 above)
Comment (mweiss17) [reply docs/models.md]: updated this

File: docs/models.md (hunk 2 above)
Comment (mweiss17) [reply docs/models.md]: updated

File: models/run.py (hunk 11 above)
Comment (mweiss17) [reply models/run.py]: made parse_args its own function -- is this what you meant? I thought it was cleaner to have it seperate (and therefore independently testable)

File: models/run.py (hunk 10 above)
Comment (mweiss17) [reply models/run.py]: ok! changed.

File: models/run.py (hunk 5 above)
Comment (mweiss17) [reply models/run.py]: ok -- changed!

File: models/risk_models.py (hunk 9 above)
Comment (mweiss17) [reply models/risk_models.py:6]: it's needed to run from root like `python models/run.py --data_path`.

We need to ship this as a python package with setup.py, which is a much better solution, but not within the scope of this PR.

File: models/risk_models.py (hunk 4 above)
Comment (mweiss17) [reply models/risk_models.py]: good call -- updated.

File: models/run.py (hunk 5 above)
Comment (mweiss17) [reply models/run.py]: updated.

File: models/run.py (hunk 6 above)
Comment (mweiss17) [reply models/run.py]: updated.

File: models/run.py (hunk 7 above)
Comment (mweiss17) [reply models/run.py]: updated.

File: models/risk_models.py (hunk 8 above)
Comment (mweiss17) [reply models/risk_models.py]: fine by me, we could also (definitely) just store this in an array in the code.

File: models/risk_models.py (hunk 9 above)
Comment (mweiss17) [reply models/risk_models.py:6]: I'm not sure how to fix this except to package the repo...

File: models/run.py (hunk 10 above)
Comment (mweiss17) [reply models/run.py]: done

File: models/run.py (hunk 11 above)
Comment (mweiss17) [reply models/run.py]: done

File: models/run.py (hunk 12 above)
Comment (mweiss17) [reply models/run.py:41]: removed dependency on args

File: models/run.py (hunk 13 above)
Comment (mweiss17) [reply models/run.py:233]: not necessary with the change I made

**Comment by mweiss17** [comment]
ok, I made the changes requested (except fixing that pathing problem that I don't know how to fix without re-packaging the repo...)

**Review by satyaog — APPROVED** [review]

**Pull request merged**